feat: Add OpenAPI support, `OpenAPITool` component

vblagoje opened the pull request:
This PR introduces the `OpenAPITool`, a powerful new component that enables calling OpenAPI service endpoints using payloads generated by chat generators from human instructions. 

### Usage Examples

**Example 1: Searching with Serper**

The following example demonstrates how to use the `OpenAPITool` with the SerperDev API to perform a search:

```python
from haystack_experimental.components.tools.openapi import OpenAPITool, LLMProvider
from haystack.dataclasses import ChatMessage
from haystack.utils import Secret

tool = OpenAPITool(generator_api=LLMProvider.OPENAI,   
                                   spec="https://bit.ly/serper_dev_spec_yaml",
                                   credentials=Secret.from_token("your_serper_key"))

results = tool.run(messages=[ChatMessage.from_user("Do a search for 'Who was Nikola Tesla?'")])
print(results)
```

**Example 2: Scraping with Firecrawl**

The following example demonstrates how to use the `OpenAPITool` with the Firecrawl API to scrape a webpage:

```python
from haystack_experimental.components.tools.openapi import OpenAPITool, LLMProvider
from haystack.dataclasses import ChatMessage
from haystack.utils import Secret

tool = OpenAPITool(generator_api=LLMProvider.OPENAI, 
                   spec="https://raw.githubusercontent.com/mendableai/firecrawl/main/apps/api/openapi.json",
                   credentials=Secret.from_env_var("FIRECRAWL_API_KEY"))

results = tool.run(messages=[ChatMessage.from_user("Scrape URL: https://news.ycombinator.com/")])
print(results)
```

**Example 3: Searching with Firecrawl using Anthropic Claude**

The following example demonstrates how to use the `OpenAPITool` with the Firecrawl API to search and get back scraped pages

```python
from haystack_experimental.components.tools.openapi import OpenAPITool, LLMProvider
from haystack.dataclasses import ChatMessage
from haystack.utils import Secret

tool = OpenAPITool(generator_api=LLMProvider.ANTHROPIC, generator_api_params={"model":"claude-3-opus-20240229"},
                   spec="https://raw.githubusercontent.com/mendableai/firecrawl/main/apps/api/openapi.json",
                   credentials=Secret.from_env_var("FIRECRAWL_API_KEY"))

results = tool.run(messages=[ChatMessage.from_user("Scrape https://haystack.deepset.ai/blog/rag-evaluation-with-prometheus-2")])
print(results)
```

## Demo

Here is a demo [notebook](https://github.com/vblagoje/notebooks/blob/main/haystack2x-demos/haystack_rag_serperdev_demo_new.ipynb) featuring RAG+SerperDev, here is [notebook](https://github.com/vblagoje/notebooks/blob/main/haystack2x-demos/haystack_anthropic_rag_firecrawl_demo.ipynb) demoing use of Anthropic models in Firecrawl WebQa RAG and finally a Cohere [notebook](https://github.com/vblagoje/notebooks/blob/main/haystack2x-demos/haystack_cohere_rag_firecrawl_demo.ipynb) variant of the same

vblagoje commented:
This is the first part of the task @shadeMe as we agreed, new tool impl comes next. 
@masci the specific filesystem layout was a consensus between two of us.

vblagoje commented:
We currently support OpenAI compatible models (OpenAI, fireworks, anyscale etc
After integration of https://github.com/deepset-ai/haystack-core-integrations/pull/790 users will be able to use Cohere and Anthropic function calling models.
More fc model support should come soon, including open source models from HF

shadeMe commented on haystack_experimental/components/tools/openapi/payload_extraction.py:
```diff
@@ -0,0 +1,80 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
```
Let's make this a private module.

shadeMe commented on haystack_experimental/components/tools/openapi/payload_extraction.py:
```diff
@@ -0,0 +1,80 @@
+    arguments_field_name: str,
+) -> Callable[[Any], Dict[str, Any]]:
+    """
+    Extracts invocation payload from a given LLM completion containing function invocation.
```
Docstring.

shadeMe commented on haystack_experimental/components/tools/openapi/schema_conversion.py:
```diff
@@ -0,0 +1,254 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
```
Private module.

shadeMe commented on haystack_experimental/components/tools/openapi/schema_conversion.py:
```diff
@@ -0,0 +1,254 @@
+    parse_endpoint_fn: Callable[[Dict[str, Any], str], Dict[str, Any]],
+) -> List[Dict[str, Any]]:
+    """
+    Extracts functions from the OpenAPI specification, converts them into a function schema.
```
Docstring

shadeMe commented on haystack_experimental/components/tools/openapi/schema_conversion.py:
```diff
@@ -0,0 +1,254 @@
+    resolved_spec: Dict[str, Any], parameters_name: str
+) -> Dict[str, Any]:
+    """
+    Parses an OpenAPI endpoint specification for OpenAI.
```
Docstring.

shadeMe commented on haystack_experimental/components/tools/openapi/schema_conversion.py:
```diff
@@ -0,0 +1,254 @@
+    property_schema: Dict[str, Any], include_attributes: Optional[List[str]] = None
+) -> Dict[str, Any]:
+    """
+    Recursively parses the attributes of a property schema.
```
Docstring (also elsewhere in this module).

TuanaCelik commented:
My first question looking at this at high level? How would I provide other haystack pipelines as tools with this component?

vblagoje commented:
> My first question looking at this at high level: How would I provide other haystack pipelines as tools with this component?

@TuanaCelik we'll have another component for that use case. This one is specific for openapi integration, hence the package name and the component name. We'll add another component in tools.pipeline or smth similar for that use case.

For this use case see https://github.com/vblagoje/notebooks/blob/main/haystack2x-demos/haystack_rag_serperdev_demo_new.ipynb

vblagoje commented:
@shadeMe I addressed 98% of the use cases as you suggested. I deviated from your advice in these situations:
- I did not lift up OpenAPISpecification to a first class citizen as it complicates use of OpenAPITool and it's DX
- I left from OpenAPISpecification.from_file (needed for testing) and I removed from_dict instead of it :-)
- I left the name credentials because although we mostly use tokens for auth, you'll see that a request for at least u/p will show up soon, and then we'll need another parameter for it breaking the OpenAPITool init
- some unintended omission might have slipped as well

vblagoje commented:
@shadeMe I added back in http bearer token auth, as this one is fairly common type of auth in addition to apiKey.  Even our own firecrawl example uses it:
```
from haystack_experimental.components.tools.openapi import OpenAPITool, LLMProvider
from haystack.dataclasses import ChatMessage

tool = OpenAPITool(generator_api=LLMProvider.ANTHROPIC, generator_api_params={"model":"claude-3-opus-20240229"},
                   tool_spec="https://raw.githubusercontent.com/mendableai/firecrawl/main/apps/api/openapi.json",
                   tool_credentials="your_firecrawl_key")

results = tool.run(messages=[ChatMessage.from_user("Search for 'What was the twitter debate between Elon Musk and Yann LeCun all about?'")])
print(results)
```

For reference OpenAPI Specification 3.x supports the following four types of authentication:

- HTTP authentication schemes (e.g., Basic, Bearer).
- API keys (sent as a query parameter, header, or cookie).
- OAuth2 (for various OAuth2 flows such as implicit, password, client credentials, and authorization code).
- OpenID Connect Discovery (for federated identity and single sign-on).

We support only first two excluding basic http (that's user password http auth)

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+            )
+        self.spec_dict = spec_dict
+
+    @classmethod
+    def from_str(cls, content: str) -> "OpenAPISpecification":
+        """
+        Create an OpenAPISpecification instance from a string.
+
+        :param content: The string content of the OpenAPI specification.
+        :return: The OpenAPISpecification instance.
+        """
+        try:
+            loaded_spec = json.loads(content)
+        except json.JSONDecodeError:
+            try:
+                loaded_spec = yaml.safe_load(content)
+            except yaml.YAMLError as e:
+                raise ValueError(
+                    "Content cannot be decoded as JSON or YAML: " + str(e)
+                ) from e
```
Let remove `from_str` for now.

vblagoje replied:
@shadeMe but this function is a must for parsing and is used from other `from_*` functions. I'll make it private

shadeMe replied:
Whoops, overlooked that entirely 🙈 Sorry, let's keep it as it is - no harm in keeping it a public function in that case. Sorry again!

vblagoje replied:
Resolved! Reverted to from_str

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+    Represents an OpenAPI specification. See https://spec.openapis.org/oas/latest.html for details.
+    """
+
+    def __init__(self, spec_dict: Dict[str, Any]):
```
Docstring

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+        Create an OpenAPISpecification instance from a file.
+
+        :param spec_file: The file path to the OpenAPI specification.
+        :return: The OpenAPISpecification instance.
```
Docstring for errors

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+        Create an OpenAPISpecification instance from a URL.
+
+        :param url: The URL to fetch the OpenAPI specification from.
+        :return: The OpenAPISpecification instance.
```
Docstring for errors

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+        :param path: The path of the operation.
+        :param path_item: The path item from the OpenAPI specification.
+        :param method: The HTTP method of the operation.
+        :return: The operation
```
Docstring for errors

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+
+    def __init__(  # noqa: PLR0913 pylint: disable=too-many-arguments
+        self,
+        openapi_spec: Union[str, Path],
```
Let's directly pass the `OpenAPISpecification`instance here instead.

vblagoje replied:
I'll do this last and it helps that `OpenAPISpecification` instance is moved into `types.py`

This is my only reservation from all of your recommendations:

- it worsens DX in OpenAPITool as we now need to first create `OpenAPISpecification` then `Configuration` then `OpenAPIServiceClient`. 
- I can do this if you insist, it causes rewrite of entire test suite and OpenAPITool

shadeMe replied:
As discussed internally, the user is in the best position to determine where the spec comes and to pick the correct intent.

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+class ClientConfiguration:
+    """Configuration for the OpenAPI client."""
+
+    def __init__(  # noqa: PLR0913 pylint: disable=too-many-arguments
```
Docstring.

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+
+        self.credentials = credentials
+        self.request_sender = request_sender or send_request
+        self.llm_provider: LLMProvider = llm_provider or LLMProvider.OPENAI
```
This should be a default value.

vblagoje replied:
@shadeMe addressed with [4f55c96](https://github.com/deepset-ai/haystack-experimental/pull/8/commits/4f55c96606af473060c6af46fb64c3537a264d7d)

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+        The function takes a security scheme and a request as arguments:
+            `security_scheme: Dict[str, Any] - The security scheme from the OpenAPI spec.`
+            `request: Dict[str, Any] - The request to apply the authentication to.`
+        :return: The authentication function.
```
Docstring for error

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+
+        :return: The tools definitions passed to the LLM as tools parameter.
+        """
+        provider_to_converter = {
+            "anthropic": anthropic_converter,
+            "cohere": cohere_converter,
+        }
```
Let's use the enum directly as the key instead of its value.

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+        This function knows how to extract the exact function payload from the LLM generated function calling payload.
+        :return: The payload extractor function.
+        """
+        provider_to_arguments_field_name = {
+            "anthropic": "input",
+            "cohere": "parameters",
+        }  # add more providers here
```
The key should be the `LLMProvider` enum.

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+                return create_api_key_auth_function(api_key=credentials)
+            if scheme["type"] == "http":
+                return create_http_auth_function(token=credentials)
+            if scheme["type"] == "oauth2":
+                raise NotImplementedError("OAuth2 authentication is not yet supported.")
```
Since we only support HTTP and API key, let's just raise for all other authentication types (and print the type in the error message).

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+
+    :param operation: The operation to build the request for.
+    :param kwargs: The arguments to use for building the request.
+    :return: The HTTP request as a dictionary.
```
Docstring for errors

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+
+    def __init__(self, client_config: ClientConfiguration):
+        self.client_config = client_config
+        self.request_sender = client_config.request_sender
```
I think we don't need to store this directly on the instance. Let's just use `self.client_config.request_sender` instead.

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+        :raises HttpClientError: If an error occurs while sending the request and receiving the response.
+        """
+        fn_extractor = self.client_config.get_payload_extractor()
+        fn_invocation_payload = fn_extractor(function_payload)
```
Don't we also need to wrap this call in `try...except`?

shadeMe commented on haystack_experimental/components/tools/openapi/_schema_conversion.py:
```diff
@@ -0,0 +1,277 @@
+logger = logging.getLogger(__name__)
+
+
+def openai_converter(schema: "OpenAPISpecification") -> List[Dict[str, Any]]:  # type: ignore[name-defined] # noqa: F821
```
I think you should be able to remove the `type: ignore` by using the following code:
```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
	# import OpenAPISpecification; won't cause circular imports
```

vblagoje replied:
@shadeMe it didn't help - I can move `Operation` and `OpenAPISpecification` to `types.py` - that works nicely and further simplifies `_openapi.py`

shadeMe commented on haystack_experimental/components/tools/openapi/_schema_conversion.py:
```diff
@@ -0,0 +1,277 @@
+
+    :param service_openapi_spec: The OpenAPI specification to extract functions from.
+    :param parameters_name: The name of the parameters field in the function schema.
+    :param parse_endpoint_fn: The function to parse the endpoint specification.
```
Missing returns

shadeMe commented on haystack_experimental/components/tools/openapi/_schema_conversion.py:
```diff
@@ -0,0 +1,277 @@
+    Parses an OpenAPI endpoint specification for OpenAI.
+
+    :param resolved_spec: The resolved OpenAPI specification.
+    :param parameters_name: The name of the parameters field in the function schema.
```
Missing returns

shadeMe commented on haystack_experimental/components/tools/openapi/_schema_conversion.py:
```diff
@@ -0,0 +1,277 @@
+    Recursively parses the attributes of a property schema.
+
+    :param property_schema: The property schema to parse.
+    :param include_attributes: The attributes to include in the parsed schema.
```
Missing returns

shadeMe commented on haystack_experimental/components/tools/openapi/_schema_conversion.py:
```diff
@@ -0,0 +1,277 @@
+    Parses an endpoint specification for Cohere.
+
+    :param operation: The operation specification to parse.
+    :param ignored_param: ignored, left for compatibility with the OpenAI converter.
```
Missing returns

shadeMe commented on haystack_experimental/components/tools/openapi/_schema_conversion.py:
```diff
@@ -0,0 +1,277 @@
+    Parses the parameters from an operation specification.
+
+    :param operation: The operation specification to parse.
+    :return: A dictionary containing the parsed parameters.
```
```suggestion
    :returns: A dictionary containing the parsed parameters.
```

shadeMe commented on haystack_experimental/components/tools/openapi/_schema_conversion.py:
```diff
@@ -0,0 +1,277 @@
+    :param schema: The schema to parse.
+    :param required: Whether the schema is required.
+    :param description: The description of the schema.
+    :return: A dictionary containing the parsed schema.
```
```suggestion
    :returns: A dictionary containing the parsed schema.
```

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,557 @@
+        )
+        if not servers:
+            raise ValueError("No servers found in the provided specification.")
+        if server_index >= len(servers):
```
We should also check that the index is not `<0`.

shadeMe commented on haystack_experimental/components/tools/openapi/types.py:
```diff
@@ -0,0 +1,14 @@
+
+class LLMProvider(Enum):
+    """
+    Enum for the supported LLM providers.
```
```suggestion
    LLM providers supported by `OpenAPITool`. 
```

shadeMe commented on haystack_experimental/components/tools/openapi/openapi_tool.py:
```diff
@@ -0,0 +1,162 @@
+        """
+        Invokes the underlying OpenAPI service/tool with the function calling payload generated by the chat generator.
+
+        :param messages: List of ChatMessages to generate function calling payload (e.g. human instructions).
```
We should mention what the expected format is seeing how we make assumptions about the list and its contents.

shadeMe commented on haystack_experimental/components/tools/openapi/openapi_tool.py:
```diff
@@ -0,0 +1,162 @@
+        generator_api: LLMProvider,
+        generator_api_params: Dict[str, Any],
+        tool_spec: Optional[Union[str, Path]] = None,
+        tool_credentials: Optional[str] = None,
```
This should be a `Secret`.

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,366 @@
+
+
+# Authentication strategies
+def create_api_key_auth_function(api_key: str):
```
Return type hint.

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,366 @@
+    return apply_auth
+
+
+def create_http_auth_function(token: str):
```
Return type hint.

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,366 @@
+        provider_to_converter = defaultdict(
+            lambda: openai_converter,
+            {
+                LLMProvider.ANTHROPIC.value: anthropic_converter,
+                LLMProvider.COHERE.value: cohere_converter,
```
We can just keep this as `Dict[LLMProvider, Callable]`:
```suggestion
                LLMProvider.ANTHROPIC: anthropic_converter,
                LLMProvider.COHERE: cohere_converter,
```

shadeMe commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,366 @@
+        provider_to_arguments_field_name = defaultdict(
+            lambda: "arguments",
+            {
+                LLMProvider.ANTHROPIC.value: "input",
+                LLMProvider.COHERE.value: "parameters",
```
Same as above.

shadeMe commented on haystack_experimental/components/tools/openapi/types.py:
```diff
@@ -0,0 +1,241 @@
+        self.spec_dict = spec_dict
+
+    @classmethod
+    def _from_str(cls, content: str) -> "OpenAPISpecification":
```
We can revert this to a public function.

vblagoje commented on haystack_experimental/components/tools/openapi/_openapi.py:
```diff
@@ -0,0 +1,341 @@
+
+    def __init__(
+        self,
+        openapi_spec: OpenAPISpecification,
```
openapi_spec is now only `OpenAPISpecification`

shadeMe commented on haystack_experimental/components/tools/openapi/openapi_tool.py:
```diff
@@ -83,7 +84,8 @@ def __init__(
             else:
                 raise ValueError(f"Invalid OpenAPI specification source {spec}. Expected valid file path or URL")
-
+            self.spec = spec  # store the spec for serialization
+            self.credentials = credentials  # store the credentials for serialization
```
This should be moved outside the `if` clause it's currently in.

shadeMe commented on haystack_experimental/components/tools/openapi/openapi_tool.py:
```diff
@@ -167,6 +169,41 @@ def run(
+            The deserialized component instance.
+        """
+        deserialize_secrets_inplace(data["init_parameters"], keys=["credentials"])
+        if "generator_api_params" in data["init_parameters"]:
```
This condition is always true, I think?

shadeMe commented on haystack_experimental/components/tools/openapi/openapi_tool.py:
```diff
@@ -167,6 +169,41 @@ def run(
+            deserialize_secrets_inplace(data["init_parameters"]["generator_api_params"], keys=["api_key"])
+        init_params = data.get("init_parameters", {})
+        generator_api = init_params.get("generator_api")
+        data["init_parameters"]["generator_api"] = LLMProvider(generator_api)
```
Let's add a `from_str` method to the `LLMProvider` enum class like we do [elsewhere](https://github.com/deepset-ai/haystack/blob/96cda5d3b6dd978f11ccbeba40ccfabf3a369042/haystack/components/extractors/named_entity_extractor.py#L39) and use it here.

shadeMe commented on test/components/tools/openapi/test_openapi_client_live.py:
```diff
@@ -30,6 +30,7 @@ def test_serperdev(self, test_files_path):
         assert "invention" in str(response)
 
     @pytest.mark.integration
+    @pytest.mark.skip("This test hits rate limit on Github API. Skip for now.")
```
Do these tests pass locally?

vblagoje replied:
Yes, but Github is a bit restrictive once it gets bombarded by all different OS/Python variants.

shadeMe commented on test/components/tools/openapi/test_openapi_tool.py:
```diff
@@ -0,0 +1,181 @@
+
+class TestOpenAPITool:
+
+    def test_to_dict(self, monkeypatch):
```
We should add another test to catch an invalid spec path/URL.

shadeMe commented on haystack_experimental/components/tools/openapi/openapi_tool.py:
```diff
@@ -167,6 +169,41 @@ def run(
+        :returns:
+            The serialized component as a dictionary.
+        """
+        if "api_key" in self.generator_api_params:
+            self.generator_api_params["api_key"] = self.generator_api_params["api_key"].to_dict()
```
This performs no validation on its assumptions. 

Looks like we don't have a `serialize_secrets_inplace` function that recursively checks values in a dict for `Secret`s and calls `to_dict` on them. Let's add that (to `haystack_experimental.util.auth`) and use that here instead.

julian-risch commented:
@vblagoje Please also add the new feature to the catalog in the readme: https://github.com/deepset-ai/haystack-experimental?tab=readme-ov-file#experiments-catalog

TuanaCelik commented:
> @vblagoje Please also add the new feature to the catalog in the readme: https://github.com/deepset-ai/haystack-experimental?tab=readme-ov-file#experiments-catalog

This made me realize that I didn't add an entry for the `OpenAIFunctionCaller` - I'll make a separate PR for it?

shadeMe reviewed (changes requested):
We just need to add the component to the readMe.

vblagoje commented:
@shadeMe Updated README to include the component 💪
